NEON: implement all intrinsics supported by architecture A64-part1

yyctw opened the pull request:
Hi all, this is Eric from Andes Technology Corporation. This PR is part of the [previous PR](https://github.com/simd-everywhere/simde/pull/1080) and includes the following:
Add 368 initial implementations and corresponding test cases in 88 families which are listed below:
- `abd`, `abdl_high`, `add`, `addhn_high`, `bsl`, `ceq`, `ceqz`, `cgez`, `cgtz`, `cle`,
- `cltz`, `cmla`, `cmla_rot180`, `cmla_rot270`, `cmla_rot90`, `cnt`, `copy_lane`, `cvt`, `cvt_n`, `cvtm`,
-  `cvtp`, `dot`, `dot_lane`, `dup_n`, `eor`, `fms_n`, `ld1`, `ld3`, `ld4`, `maxnm`,
-  `maxv`, `minnm`, `minv`, `mull`, `mull_high`, `mvn`, `pmin`, `qrdmulh_lane`, `qrshl`, `qrshrn_high_n`,
-  `qrshrun_high_n`, `qshl_n`, `qshlu_n`, `qshrn_high_n`, `qshrn_n`, `qshrun_n`, `qtbl`, `qtbx`, `raddhn`, `raddhn_high`,
-  `rbit`, `reinterpret`, `rev16`, `rev32`, `rev64`, `rnd`, `rndi`, `rndm`, `rndp`, `rshrn_high_n`,
-  `rsubhn`, `rsubhn_high`, `shr_n`, `shrn_n`, `sli_n`, `sri_n`, `st1`, `st1_lane`, `st1_x2`, `st1_x3`,
-  `st1_x4`, `st1q_x2`, `st1q_x3`, `st1q_x4`, `st2_lane`, `st3`, `st3_lane`, `st4`, `st4_lane`, `tbl`,
-  `tbx`, `trn`, `trn1`, `trn2`, `tst`, `uzp`, `uzp1`, `uzp2`

mr-c commented:
The build error for "Windows MSVC-Aarch64" reminds me of this note in the `.appveyor.yml`.

https://github.com/simd-everywhere/simde/blob/029d7498974eb43fe98b0341ca1dd269ca70d12c/.appveyor.yml#L99

So maybe try replacing 
https://github.com/simd-everywhere/simde/blob/029d7498974eb43fe98b0341ca1dd269ca70d12c/.github/workflows/ci.yml#L662
with `ninja -C build test`

mr-c commented on simde/arm/neon/rsubhn_high.h:
```diff
@@ -0,0 +1,102 @@
+#endif
+#if defined(SIMDE_ARM_NEON_A64V8_ENABLE_NATIVE_ALIASES)
+  #undef vrsubhn_high_s16
+  #define vrsubhn_high_s16(a, b) simde_vrsubhn_high_s16((a), (b))
```
```suggestion
  #define vrsubhn_high_s16(r, a, b) simde_vrsubhn_high_s16((r), (a), (b))
```

mr-c commented on simde/arm/neon/rsubhn_high.h:
```diff
@@ -0,0 +1,102 @@
+#endif
+#if defined(SIMDE_ARM_NEON_A64V8_ENABLE_NATIVE_ALIASES)
+  #undef vrsubhn_high_s32
+  #define vrsubhn_high_s32(a, b) simde_vrsubhn_high_s32((a), (b))
```
```suggestion
  #define vrsubhn_high_s32(r, a, b) simde_vrsubhn_high_s32((r), (a), (b))
```

mr-c commented on simde/arm/neon/rsubhn_high.h:
```diff
@@ -0,0 +1,102 @@
+#endif
+#if defined(SIMDE_ARM_NEON_A64V8_ENABLE_NATIVE_ALIASES)
+  #undef vrsubhn_high_s64
+  #define vrsubhn_high_s64(a, b) simde_vrsubhn_high_s64((a), (b))
```
```suggestion
  #define vrsubhn_high_s64(r, a, b) simde_vrsubhn_high_s64((r), (a), (b))
```

mr-c commented on simde/arm/neon/rsubhn_high.h:
```diff
@@ -0,0 +1,102 @@
+#endif
+#if defined(SIMDE_ARM_NEON_A64V8_ENABLE_NATIVE_ALIASES)
+  #undef vrsubhn_high_u16
+  #define vrsubhn_high_u16(a, b) simde_vrsubhn_high_u16((a), (b))
```
```suggestion
  #define vrsubhn_high_u16(r, a, b) simde_vrsubhn_high_u16((r), (a), (b))
```

mr-c commented on simde/arm/neon/rsubhn_high.h:
```diff
@@ -0,0 +1,102 @@
+#endif
+#if defined(SIMDE_ARM_NEON_A64V8_ENABLE_NATIVE_ALIASES)
+  #undef vrsubhn_high_u32
+  #define vrsubhn_high_u32(a, b) simde_vrsubhn_high_u32((a), (b))
```
```suggestion
  #define vrsubhn_high_u32(r, a, b) simde_vrsubhn_high_u32((r), (a), (b))
```

mr-c commented on simde/arm/neon/rsubhn_high.h:
```diff
@@ -0,0 +1,102 @@
+#endif
+#if defined(SIMDE_ARM_NEON_A64V8_ENABLE_NATIVE_ALIASES)
+  #undef vrsubhn_high_u64
+  #define vrsubhn_high_u64(a, b) simde_vrsubhn_high_u64((a), (b))
```
```suggestion
  #define vrsubhn_high_u64(r, a, b) simde_vrsubhn_high_u64((r), (a), (b))
```

mr-c reviewed (changes requested):
Quick fix for the native-aliases build failure

yyctw commented:
> Quick fix for the native-aliases build failure

It has been fixed.

mr-c commented on simde/arm/neon/add.h:
```diff
@@ -22,6 +22,7 @@
  *
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

Not yet, I don't think

mr-c commented on simde/arm/neon/bsl.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
  *   2020      Sean Maher <seanptmaher@gmail.com> (Copyright owned by Google, LLC)
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/ceq.h:
```diff
@@ -21,7 +21,7 @@
  *
  * Copyright:
- *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
*   2020      Evan Nemerson <evan@nemerson.com>
```
Evan wrote the first version of this file..

mr-c commented on simde/arm/neon/ceqz.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
  *   2020      Sean Maher <seanptmaher@gmail.com> (Copyright owned by Google, LLC)
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/cmla.h:
```diff
@@ -22,6 +22,7 @@
 *
 * Copyright:
 *   2021      Atharva Nimbalkar <atharvakn@gmail.com>
+*   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/cmla_rot180.h:
```diff
@@ -22,6 +22,7 @@
 *
 * Copyright:
 *   2021      Atharva Nimbalkar <atharvakn@gmail.com>
+*   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/cmla_rot270.h:
```diff
@@ -22,6 +22,7 @@
 *
 * Copyright:
 *   2021      Atharva Nimbalkar <atharvakn@gmail.com>
+*   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/cmla_rot90.h:
```diff
@@ -22,6 +22,7 @@
 *
 * Copyright:
 *   2021      Atharva Nimbalkar <atharvakn@gmail.com>
+*   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/cnt.h:
```diff
@@ -21,7 +21,7 @@
  *
  * Copyright:
- *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
 *   2020      Evan Nemerson <evan@nemerson.com>
```

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -144,11 +144,14 @@ int16_t
 simde_vcvth_s16_f16(simde_float16 a) {
   #if defined(SIMDE_ARM_NEON_A64V8_NATIVE) && defined(SIMDE_ARM_NEON_FP16)
     return vcvth_s16_f16(a);
+  #elif defined(SIMDE_FAST_CONVERSION_RANGE)
+    return HEDLEY_STATIC_CAST(int16_t,
+        simde_float16_to_float32(a));
```
There should be test cases added to `test_simde_vcvth_s16_f16` that show the different behavior if `SIMDE_FAST_CONVERSION_RANGE` is defined, or not.

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -144,11 +144,14 @@ int16_t
+    return HEDLEY_STATIC_CAST(int16_t,
+        simde_float16_to_float32(a));
   #else
     simde_float32 af = simde_float16_to_float32(a);
-    if (HEDLEY_UNLIKELY(af < HEDLEY_STATIC_CAST(simde_float32, INT16_MIN))) {
+    if (HEDLEY_UNLIKELY(af <= HEDLEY_STATIC_CAST(simde_float32, INT16_MIN))) {
       return INT16_MIN;
-    } else if (HEDLEY_UNLIKELY(af > HEDLEY_STATIC_CAST(simde_float32, INT16_MAX))) {
+    } else if (HEDLEY_UNLIKELY(af >= HEDLEY_STATIC_CAST(simde_float32, INT16_MAX))) {
       return INT16_MAX;
```
Likewise, there should be test cases in `test_simde_vcvth_s16_f16` that trigger these code paths involving `INT16_{MIN,MAX}`

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -185,6 +191,110 @@ simde_vcvth_u16_f16(simde_float16 a) {
+simde_vcvth_s32_f16(simde_float16 a) {
+  #if defined(SIMDE_ARM_NEON_A32V8_NATIVE) && defined(SIMDE_ARM_NEON_FP16)
+    return vcvth_s32_f16(a);
+  #elif defined(SIMDE_FAST_CONVERSION_RANGE)
+    return HEDLEY_STATIC_CAST(int32_t,
+        simde_float16_to_float32(a));
+  #else
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af <= HEDLEY_STATIC_CAST(simde_float32, INT32_MIN))) {
+      return INT32_MIN;
+    } else if (HEDLEY_UNLIKELY(af >= HEDLEY_STATIC_CAST(simde_float32, INT32_MAX))) {
+      return INT32_MAX;
+    } else if (HEDLEY_UNLIKELY(simde_math_isnan(af))) {
+      return 0;
```
Same comments as for `simde_vcvth_s16_f16`

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -144,11 +144,14 @@ int16_t
-    } else if (HEDLEY_UNLIKELY(af > HEDLEY_STATIC_CAST(simde_float32, INT16_MAX))) {
+    } else if (HEDLEY_UNLIKELY(af >= HEDLEY_STATIC_CAST(simde_float32, INT16_MAX))) {
       return INT16_MAX;
     } else if (HEDLEY_UNLIKELY(simde_math_isnanf(af))) {
       return 0;
```
And a test case involving a NAN; probably better to test with `simde_math_isnanhf(a)` as we might get an optimized implementation someday

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -167,11 +170,14 @@ uint16_t
     return vcvth_u16_f16(a);
+  #elif defined(SIMDE_FAST_CONVERSION_RANGE)
+    return HEDLEY_STATIC_CAST(uint16_t,
+        simde_float16_to_float32(a));
   #else
     simde_float32 af = simde_float16_to_float32(a);
-    if (HEDLEY_UNLIKELY(af < SIMDE_FLOAT32_C(0.0))) {
+    if (HEDLEY_UNLIKELY(af <= SIMDE_FLOAT32_C(0.0))) {
       return 0;
-    } else if (HEDLEY_UNLIKELY(af > HEDLEY_STATIC_CAST(simde_float32, UINT16_MAX))) {
+    } else if (HEDLEY_UNLIKELY(af >= HEDLEY_STATIC_CAST(simde_float32, UINT16_MAX))) {
       return UINT16_MAX;
     } else if (simde_math_isnanf(af)) {
       return 0;
```
Same comments as for `simde_vcvth_s16_f16`

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -185,6 +191,110 @@ simde_vcvth_u16_f16(simde_float16 a) {
+simde_vcvth_u32_f16(simde_float16 a) {
+  #if defined(SIMDE_ARM_NEON_A32V8_NATIVE) && defined(SIMDE_ARM_NEON_FP16)
+    return vcvth_u32_f16(a);
+  #elif defined(SIMDE_FAST_CONVERSION_RANGE)
+    return HEDLEY_STATIC_CAST(uint32_t,
+        simde_float16_to_float32(a));
+  #else
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af <= HEDLEY_STATIC_CAST(simde_float32, 0.0))) {
+      return 0;
+    } else if (HEDLEY_UNLIKELY(af >= HEDLEY_STATIC_CAST(simde_float32, UINT32_MAX))) {
+      return UINT32_MAX;
+    } else if (HEDLEY_UNLIKELY(simde_math_isnanf(af))) {
+      return 0;
```
Same comments as for `simde_vcvth_s16_f16`

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -185,6 +191,110 @@ simde_vcvth_u16_f16(simde_float16 a) {
+simde_vcvth_s64_f16(simde_float16 a) {
+  #if defined(SIMDE_ARM_NEON_A64V8_NATIVE) && defined(SIMDE_ARM_NEON_FP16)
+    return vcvth_s64_f16(a);
+  #elif defined(SIMDE_FAST_CONVERSION_RANGE)
+    return HEDLEY_STATIC_CAST(int64_t,
+        simde_float16_to_float32(a));
+  #else
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af <= HEDLEY_STATIC_CAST(simde_float32, INT64_MIN))) {
+      return INT64_MIN;
+    } else if (HEDLEY_UNLIKELY(af >= HEDLEY_STATIC_CAST(simde_float32, INT64_MAX))) {
+      return INT64_MAX;
+    } else if (HEDLEY_UNLIKELY(simde_math_isnanf(af))) {
+      return 0;
```
Same comments as for `simde_vcvth_s16_f16`

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -185,6 +191,110 @@ simde_vcvth_u16_f16(simde_float16 a) {
+simde_vcvth_u64_f16(simde_float16 a) {
+  #if defined(SIMDE_ARM_NEON_A64V8_NATIVE) && defined(SIMDE_ARM_NEON_FP16)
+    return vcvth_u64_f16(a);
+  #elif defined(SIMDE_FAST_CONVERSION_RANGE)
+    return HEDLEY_STATIC_CAST(uint64_t,
+        simde_float16_to_float32(a));
+  #else
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af <= HEDLEY_STATIC_CAST(simde_float32, 0.0))) {
+      return 0;
+    } else if (HEDLEY_UNLIKELY(af >= HEDLEY_STATIC_CAST(simde_float32, UINT64_MAX))) {
+      return UINT64_MAX;
+    } else if (HEDLEY_UNLIKELY(simde_math_isnanf(af))) {
+      return 0;
```
Same comments as for `simde_vcvth_s16_f16`

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1177,11 +1383,199 @@ simde_vcvtq_f64_u64(simde_uint64x2_t a) {
+    return vcvtah_s16_f16(a);
+  #else
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af < HEDLEY_STATIC_CAST(simde_float32, INT16_MIN))) {
+      return INT16_MIN;
+    } else if (HEDLEY_UNLIKELY(af > HEDLEY_STATIC_CAST(simde_float32, INT16_MAX))) {
+      return INT16_MAX;
+    } else if (HEDLEY_UNLIKELY(simde_math_isnanf(af))) {
+      return 0;
```
Needs test coverage

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -185,6 +191,110 @@ simde_vcvth_u16_f16(simde_float16 a) {
+        simde_float16_to_float32(a));
+  #else
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af <= HEDLEY_STATIC_CAST(simde_float32, 0.0))) {
```
```suggestion
    if (HEDLEY_UNLIKELY(af <= SIMDE_FLOAT32_C(0.0))) {
```

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -185,6 +191,110 @@ simde_vcvth_u16_f16(simde_float16 a) {
+        simde_float16_to_float32(a));
+  #else
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af <= HEDLEY_STATIC_CAST(simde_float32, 0.0))) {
```
```suggestion
    if (HEDLEY_UNLIKELY(af <= SIMDE_FLOAT32_C(0.0))) {
```

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1177,11 +1383,199 @@ simde_vcvtq_f64_u64(simde_uint64x2_t a) {
+    return vcvtah_u16_f16(a);
+  #else
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af < HEDLEY_STATIC_CAST(simde_float32, 0.0))) {
```
```suggestion
    if (HEDLEY_UNLIKELY(af < SIMDE_FLOAT32_C(0.0))) {
```

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1177,11 +1383,199 @@ simde_vcvtq_f64_u64(simde_uint64x2_t a) {
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af < HEDLEY_STATIC_CAST(simde_float32, 0.0))) {
+      return 0;
+    } else if (HEDLEY_UNLIKELY(af > HEDLEY_STATIC_CAST(simde_float32, UINT16_MAX))) {
+      return UINT16_MAX;
+    } else if (HEDLEY_UNLIKELY(simde_math_isnanf(af))) {
+      return 0;
```
Needs test coverage

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1177,11 +1383,199 @@ simde_vcvtq_f64_u64(simde_uint64x2_t a) {
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af < HEDLEY_STATIC_CAST(simde_float32, INT32_MIN))) {
+      return INT32_MIN;
+    } else if (HEDLEY_UNLIKELY(af > HEDLEY_STATIC_CAST(simde_float32, INT32_MAX))) {
+      return INT32_MAX;
+    } else if (HEDLEY_UNLIKELY(simde_math_isnanf(af))) {
+      return 0;
```
Needs test coverage

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1177,11 +1383,199 @@ simde_vcvtq_f64_u64(simde_uint64x2_t a) {
+    return vcvtah_u32_f16(a);
+  #else
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af < HEDLEY_STATIC_CAST(simde_float32, 0.0))) {
```
```suggestion
    if (HEDLEY_UNLIKELY(af < SIMDE_FLOAT32_C(0.0))) {
```

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1177,11 +1383,199 @@ simde_vcvtq_f64_u64(simde_uint64x2_t a) {
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af < HEDLEY_STATIC_CAST(simde_float32, 0.0))) {
+      return 0;
+    } else if (HEDLEY_UNLIKELY(af > HEDLEY_STATIC_CAST(simde_float32, UINT32_MAX))) {
+      return UINT32_MAX;
+    } else if (HEDLEY_UNLIKELY(simde_math_isnanf(af))) {
+      return 0;
```
Needs test coverage

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1177,11 +1383,199 @@ simde_vcvtq_f64_u64(simde_uint64x2_t a) {
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af < HEDLEY_STATIC_CAST(simde_float32, INT64_MIN))) {
+      return INT64_MIN;
+    } else if (HEDLEY_UNLIKELY(af > HEDLEY_STATIC_CAST(simde_float32, INT64_MAX))) {
+      return INT64_MAX;
+    } else if (HEDLEY_UNLIKELY(simde_math_isnanf(af))) {
+      return 0;
```
Needs test coverage

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1177,11 +1383,199 @@ simde_vcvtq_f64_u64(simde_uint64x2_t a) {
+    return vcvtah_u64_f16(a);
+  #else
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af < HEDLEY_STATIC_CAST(simde_float32, 0.0))) {
```
```suggestion
    if (HEDLEY_UNLIKELY(af < SIMDE_FLOAT32_C(0.0))) {
```

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1177,11 +1383,199 @@ simde_vcvtq_f64_u64(simde_uint64x2_t a) {
+    simde_float32 af = simde_float16_to_float32(a);
+    if (HEDLEY_UNLIKELY(af < HEDLEY_STATIC_CAST(simde_float32, 0.0))) {
+      return 0;
+    } else if (HEDLEY_UNLIKELY(af > HEDLEY_STATIC_CAST(simde_float32, UINT64_MAX))) {
+      return UINT64_MAX;
+    } else if (HEDLEY_UNLIKELY(simde_math_isnanf(af))) {
+      return 0;
```
Needs test coverage

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1177,11 +1383,199 @@ simde_vcvtq_f64_u64(simde_uint64x2_t a) {
+simde_vcvtad_s64_f64(simde_float64 a) {
+  #if defined(SIMDE_ARM_NEON_A64V8_NATIVE)
+    return vcvtad_s64_f64(a);
+  #elif defined(SIMDE_FAST_CONVERSION_RANGE)
+    return HEDLEY_STATIC_CAST(int64_t, simde_math_round(a));
+  #else
+    if (HEDLEY_UNLIKELY(a <= HEDLEY_STATIC_CAST(simde_float64, INT64_MIN))) {
+      return INT64_MIN;
+    } else if (HEDLEY_UNLIKELY(a >= HEDLEY_STATIC_CAST(simde_float64, INT64_MAX))) {
+      return INT64_MAX;
+    } else if (HEDLEY_UNLIKELY(simde_math_isnan(a))) {
+      return 0;
```
Needs test coverage

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1177,11 +1383,199 @@ simde_vcvtq_f64_u64(simde_uint64x2_t a) {
+simde_vcvtad_u64_f64(simde_float64 a) {
+  #if defined(SIMDE_ARM_NEON_A64V8_NATIVE) && !defined(SIMDE_BUG_CLANG_46844)
+    return vcvtad_u64_f64(a);
+  #elif defined(SIMDE_FAST_CONVERSION_RANGE)
+    return HEDLEY_STATIC_CAST(uint64_t, simde_math_round(a));
+  #else
+    if (HEDLEY_UNLIKELY(a <= SIMDE_FLOAT64_C(0.0))) {
+      return 0;
+    } else if (HEDLEY_UNLIKELY(a >= HEDLEY_STATIC_CAST(simde_float64, UINT64_MAX))) {
+      return UINT64_MAX;
+    } else if (simde_math_isnan(a)) {
+      return 0;
```
Needs test coverage

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1177,11 +1383,199 @@ simde_vcvtq_f64_u64(simde_uint64x2_t a) {
 simde_vcvtas_s32_f32(simde_float32 a) {
   #if defined(SIMDE_ARM_NEON_A64V8_NATIVE)
     return vcvtas_s32_f32(a);
+  #elif defined(SIMDE_FAST_CONVERSION_RANGE)
+    return HEDLEY_STATIC_CAST(int32_t, simde_math_roundf(a));
```
Needs test coverage

mr-c commented on simde/arm/neon/cvt.h:
```diff
@@ -1204,6 +1598,8 @@ uint32_t
 simde_vcvtas_u32_f32(simde_float32 a) {
   #if defined(SIMDE_ARM_NEON_A64V8_NATIVE)
     return vcvtas_u32_f32(a);
+  #elif defined(SIMDE_FAST_CONVERSION_RANGE)
+    return HEDLEY_STATIC_CAST(uint32_t, simde_math_roundf(a));
```
Needs test coverage

mr-c commented on simde/arm/neon/cvtm.h:
same comments as `cvt.h`

mr-c commented on simde/arm/neon/cvtp.h:
same comments as `cvt.h`

mr-c commented on simde/arm/neon/dot.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
  *   2020      Sean Maher <seanptmaher@gmail.com> (Copyright owned by Google, LLC)
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/dot_lane.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
  *   2020      Sean Maher <seanptmaher@gmail.com> (Copyright owned by Google, LLC)
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/dup_n.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2020      Sean Maher <seanptmaher@gmail.com> (Copyright owned by Google, LLC)
  *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/eor.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
  *   2020      Christopher Moore <moore@free.fr>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/ld1.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
  *   2021      Zhi An Ng <zhin@google.com> (Copyright owned by Google, LLC)
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/maxnm.h:
```diff
@@ -22,6 +22,7 @@
  *
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/maxv.h:
```diff
@@ -22,6 +22,7 @@
  *
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/minnm.h:
```diff
@@ -22,6 +22,7 @@
  *
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/minv.h:
```diff
@@ -22,6 +22,7 @@
  *
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/mull.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
  *   2020      Sean Maher <seanptmaher@gmail.com> (Copyright owned by Google, LLC)
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/mull_high.h:
```diff
@@ -22,6 +22,7 @@
  *
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/mvn.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
  *   2020      Christopher Moore <moore@free.fr>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/pmin.h:
```diff
@@ -22,6 +22,7 @@
  *
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/qrdmulh_lane.h:
```diff
@@ -22,6 +22,7 @@
  *
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/qrshl.h:
```diff
@@ -0,0 +1,770 @@
+#include "../../x86/avx.h"
+#include "types.h"
+
+/* Notes from the implementer (Christopher Moore aka rosbif)
```
Does this note still make sense here?

mr-c replied:
Maybe better to refer to the original file instead of duplicating the comment

mr-c commented on simde/arm/neon/qshlu_n.h:
```diff
@@ -22,6 +22,7 @@
  *
  * Copyright:
  *   2021      Atharva Nimbalkar <atharvakn@gmail.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/qshrun_n.h:
```diff
@@ -22,6 +22,7 @@
  *
  * Copyright:
  *   2021      Zhi An Ng <zhin@google.com> (Copyright owned by Google, LLC)
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/qtbl.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
  *   2020      Christopher Moore <moore@free.fr>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/qtbx.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
  *   2020      Christopher Moore <moore@free.fr>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c reviewed (changes requested):
Initial review of the 1st 50 files; I'll review the next batch when these comments are resolved.

yyctw commented:
> Initial review of the 1st 50 files; I'll review the next batch when these comments are resolved.

@mr-c I have addressed all the comments. My apologies for inadvertently deleting and adding the Copyright. Thank you for your review!

mr-c commented on simde/arm/neon/sri_n.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2021      Zhi An Ng <zhin@google.com> (Copyright owned by Google, LLC)
  *   2021      Evan Nemerson <evan@nemerson.com>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c commented on simde/arm/neon/st1_x2.h:
```diff
@@ -21,8 +21,6 @@
  * SOFTWARE.
  *
  * Copyright:
- *   2020      Evan Nemerson <evan@nemerson.com>
- *   2021      Décio Luiz Gazzoni Filho <decio@decpp.net>
```
These shouldn't be deleted

mr-c commented on simde/arm/neon/st1_x3.h:
```diff
@@ -21,8 +21,6 @@
  * SOFTWARE.
  *
  * Copyright:
- *   2020      Evan Nemerson <evan@nemerson.com>
- *   2021      Décio Luiz Gazzoni Filho <decio@decpp.net>
```
.

mr-c commented on simde/arm/neon/st1_x4.h:
```diff
@@ -21,8 +21,6 @@
  * SOFTWARE.
  *
  * Copyright:
- *   2020      Evan Nemerson <evan@nemerson.com>
- *   2021      Décio Luiz Gazzoni Filho <decio@decpp.net>
```
.

mr-c commented on simde/arm/neon/st1q_x4.h:
```diff
@@ -21,8 +21,6 @@
  * SOFTWARE.
  *
  * Copyright:
- *   2020      Evan Nemerson <evan@nemerson.com>
- *   2021      Décio Luiz Gazzoni Filho <decio@decpp.net>
```
.

mr-c commented on simde/arm/neon/tbl.h:
```diff
@@ -23,6 +23,7 @@
  * Copyright:
  *   2020      Evan Nemerson <evan@nemerson.com>
  *   2020      Christopher Moore <moore@free.fr>
+ *   2023      Yi-Yen Chung <eric681@andestech.com> (Copyright owned by Andes Technology)
```
```suggestion
```

mr-c reviewed (changes requested):
1st 75 files reviewed. Some errant copyright statements still.

Thank you for including the test generation code!

yyctw commented:
> 1st 75 files reviewed. Some errant copyright statements still.
> 
> Thank you for including the test generation code!

Already fixed.

mr-c reviewed (approved):
LGTM!
